### PR TITLE
Bug fix for multithreaded rendering.

### PR DESCRIPTION
**Pull request by wojzaremba**

### CODE REVIEW

File: mujoco_py/mjsim.pyx [hunk 1]
```diff
@@ -1,5 +1,8 @@
 from xml.dom import minidom
 from mujoco_py.utils import remove_empty_lines
+from threading import Lock
+
+render_lock = Lock()
```
Comment (welinder) [inline mujoco_py/mjsim.pyx]: this name is too generic to be in the global namespace. Let's call it `_MjSim_render_lock`

File: mujoco_py/mjsim.pyx [hunk 2]
```diff
@@ -122,16 +125,17 @@ cdef class MjSim(object):
...
-                render_context = MjRenderContextOffscreen(
-                    self, device_id=device_id)
-            else:
-                render_context = self._render_context_offscreen
-
-            render_context.render(
-                width=width, height=height, camera_id=camera_id)
-            return render_context.read_pixels(
-                width, height, depth=depth)
+            with render_lock:
```
Comment (welinder) [inline mujoco_py/mjsim.pyx]: you should be able to move this down to L134

File: mujoco_py/tests/test_cymj.py [hunk 3]
```diff
@@ -651,6 +650,33 @@ def test_sensors():
 
 
 @requires_rendering
+def test_lock():
```
Comment (welinder) [inline mujoco_py/tests/test_cymj.py]: let's call it `test_concurrent_rendering`

File: mujoco_py/tests/test_cymj.py [hunk 4]
```diff
@@ -651,6 +650,33 @@ def test_sensors():
...
+    # Tests whetever multithreaded rendering is safe.
```
Comment (welinder) [inline mujoco_py/tests/test_cymj.py]: How about this comment:

> Best-effort testing that concurrent multi-threaded rendering works. The test has no guarantees around being deterministic, but if it fails you know something is wrong with concurrent rendering. If it passes, things are *probably* working.

File: mujoco_py/tests/test_cymj.py [hunk 5]
```diff
@@ -651,6 +650,33 @@ def test_sensors():
...
+    model = load_model_from_xml(BASIC_MODEL_XML)
+    sim = MjSim(model)
+    sim.render(100, 100)
+    def func():
+        nonlocal sim
+        sim.data.qpos[:] = 0.0
+        sim.forward()
+        img1 = sim.render(width=40, height=40, camera_name="camera1")
```
Comment (welinder) [inline mujoco_py/tests/test_cymj.py]: You could use a shared `threading.Event` variable here: let all the threads wait for it, then set it to make sure more of the threads run concurrently. In the current implementation, there's always the chance that one thread finishes before the next one starts. Using an event reduces (but doesn't eliminate) it.

If you do this, you could also also do `thread.start()` right after instantiation.

File: mujoco_py/tests/test_cymj.py (hunk 4 above)
Comment (wojzaremba) [inline mujoco_py/tests/test_cymj.py]: done

File: mujoco_py/mjsim.pyx (hunk 1 above)
Comment (wojzaremba) [inline mujoco_py/mjsim.pyx]: done

File: mujoco_py/tests/test_cymj.py (hunk 3 above)
Comment (wojzaremba) [inline mujoco_py/tests/test_cymj.py]: done

File: mujoco_py/mjsim.pyx (hunk 2 above)
Comment (wojzaremba) [inline mujoco_py/mjsim.pyx]: cann't we get race condition in adding context to render_context list ?

**Pull request merged**